### PR TITLE
Grype

### PR DESCRIPTION
**Pull request by akpsgit**

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -33,31 +30,17 @@ const (
...
-				correct = true
-				break
-			}
-		}
-
-		if !correct {
-			return "", fmt.Errorf("Clair output level %s is not supported, only support %v\n", outputEnv, Priorities)
-		}
+func validateThresholdSeverity(severity string) (error) {
+	if !slice.ContainsString(severities, severity, nil) {
```
Comment (FrimIdan) [inline config/config.go]: should it be a case sensitive test?

File: config/config.go [hunk 2]
```diff
@@ -78,34 +61,34 @@ func parseBoolOption(key string) bool {
...
 	WhiteListFile     string
 	IgnoreUnfixed     bool
 	ResultServicePath string
 }
 
 func NewConfig(imageName string) (*Config, error) {
-	clairAddr := os.Getenv(OptionClairAddress)
-	if clairAddr == "" {
-		return nil, fmt.Errorf("clair address must be provided")
+	grypeAddr := os.Getenv(OptionGrypeAddress)
```
Comment (FrimIdan) [inline config/config.go:62]: we should move to `viper` here

File: config/config.go [hunk 3]
```diff
@@ -78,34 +61,34 @@ func parseBoolOption(key string) bool {
...
+	severityThreshold := os.Getenv(OptionSeverityThreshold)
+	err := validateThresholdSeverity(severityThreshold)
 	if err != nil {
 		return nil, err
 	}
 
-	clairTimeout := parseIntOption(OptionClairTimeout)
-	if clairTimeout == 0 {
-		clairTimeout = 1
+	scanTimeout := parseIntOption(OptionScanTimeout)
```
Comment (FrimIdan) [inline config/config.go]: I'm not sure that `scanTimeout` is used.

File: docker/docker.go [hunk 4]
```diff
@@ -363,13 +363,13 @@ func FetchFsCommands(config *Config) ([]*FsLayerCommand, error) {
 
 	var layerCommands []*FsLayerCommand
 	for i, layer := range layers {
-		layerDigest, err := layer.Digest()
+		layerDiffID, err := layer.DiffID()
```
Comment (FrimIdan) [inline docker/docker.go]: maybe need to comment here why the `Hash of the uncompressed layer` is used?

File: docker/token/secret/secret_test.go [hunk 5]
```diff
@@ -51,18 +51,18 @@ func TestImagePullSecret_GetCredentials(t *testing.T) {
...
+		//	fields: fields{
+		//		body: "{\"auths\":{\"gcr.io\":{\"username\":\"gcr\",\"password\":\"io\",\"auth\":\"Z2NyOmlv\"},\"gcr.io/more/specific\":{\"username\":\"gcr\",\"password\":\"io/more/specific\",\"auth\":\"Z2NyOmlvL21vcmUvc3BlY2lmaWM=\"},\"http://foo.example.com\":{\"username\":\"foo\",\"password\":\"bar\",\"auth\":\"Zm9vOmJhcg==\"}}}",
+		//	},
+		//	args: args{
+		//		named: noMatch,
+		//	},
+		//	wantUsername: "",
+		//	wantPassword: "",
+		//	wantErr:      true,
+		//},
```
Comment (FrimIdan) [inline docker/token/secret/secret_test.go]: not sure what happen - but it fails also in my local env with master.
for some reason `dockerKeyring.Lookup(named.Name())` returns credentials list with empty username and password.
I think it related to the fact that this is a docker.io example image - when changing the image to `reference.ParseNormalizedNamed("no.match.com/no/match:123")` test passed.

we can also add 
```

	if credentials[0].Username == "" || credentials[0].Password == "" {
		return "", "", fmt.Errorf("failed to get image credentials. image=%v", named.Name())
	}
```

after 
```
	credentials, credentialsExist := dockerKeyring.Lookup(named.Name())
	if !credentialsExist {
		return "", "", fmt.Errorf("failed to get image credentials. image=%v", named.Name())
	}
```

File: go.mod [hunk 6]
```diff
@@ -1,21 +1,28 @@
...
 	github.com/hashicorp/go-multierror v1.1.0
-	github.com/sirupsen/logrus v1.6.0
+	github.com/sirupsen/logrus v1.8.1
 	golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543
 	k8s.io/api v0.17.4
 	k8s.io/kubernetes v1.17.3
 )
 
+//replace github.com/Portshift/grype-server/api v0.0.0 => /Users/alexeik/go/src/github.com/Portshift/grype-server/api
+
```
Comment (FrimIdan) [inline go.mod]: remove

File: main.go [hunk 7]
```diff
@@ -81,18 +76,20 @@ func initLogs() {
...
 
-func filterVulnerabilities(severityThresholdStr string, vulnerabilities []*clair.Vulnerability) []*clair.Vulnerability {
-	var ret []*clair.Vulnerability
+func filterVulnerabilities(severityThresholdStr string, vulnerabilities *grype_models.Document) *grype_models.Document{
+	var filtered []grype_models.Match
 
 	severityThreshold := vulutils.GetSeverityFromString(severityThresholdStr)
-	for _, vulnerability := range vulnerabilities {
-		if vulutils.GetSeverityFromString(vulnerability.Severity) < severityThreshold {
+	for _, vulnerability := range vulnerabilities.Matches {
```
Comment (FrimIdan) [inline main.go:83]: since `vulnerability` is a tmp copy It will be better to use `vulnerabilities.Matches[i]` instead.

File: run/run.go [hunk 8]
```diff
@@ -1,47 +1,154 @@
...
 
-func ExecuteScan(conf *config.Config) ([]*clair.Vulnerability, []*docker.FsLayerCommand, error) {
-	image, err := docker.NewImage(&conf.DockerConfig)
+func ExecuteRemoteGrypeScan(imageName string, conf *config.Config) (*grype_models.Document, []*docker.FsLayerCommand, error) {
+	src, cleanup, err := source.New(imageName, &anchore_image.RegistryOptions{
+		InsecureSkipTLSVerify: conf.DockerConfig.InsecureTLS,
+		InsecureUseHTTP:       conf.DockerConfig.InsecureRegistry,
+		Credentials: []anchore_image.RegistryCredentials{
+			{
+				Authority: "", // What is this?
```
Comment (FrimIdan) [inline run/run.go]: from syft README
```
  # credentials for specific registries
  auth:
    - # the URL to the registry (e.g. "docker.io", "localhost:5000", etc.)
      # SYFT_REGISTRY_AUTH_AUTHORITY env var
      authority: ""
```

not sure what are the implication if it's empty.

File: run/run.go [hunk 9]
```diff
@@ -1,47 +1,154 @@
...
+				Username:  conf.DockerConfig.User,
+				Password:  conf.DockerConfig.Password,
+				Token:     conf.DockerConfig.Token,
```
Comment (FrimIdan) [inline run/run.go]: did you tested it for private registries?

File: run/run.go [hunk 10]
```diff
@@ -1,47 +1,154 @@
...
 	}
 
-	if err := image.FetchFsCommands(&conf.DockerConfig); err != nil {
-		return nil, nil, fmt.Errorf("failed to fetch layer commands: %v", err)
+	sbom64 := base64.StdEncoding.EncodeToString(sbomEncoded)
+
+	client := createGrypeClient(conf.GrypeAddr)
+	params := grype_client_operations.NewPostScanSBOMParams().WithBody(&models.SBOM{
+		Sbom: sbom64,
+	})
```
Comment (FrimIdan) [inline run/run.go]: need to add `WithContext` or `WithTimeout` to support scan timeout

File: run/run.go [hunk 11]
```diff
@@ -1,47 +1,154 @@
...
+	}
+	doc := grype_models.Document{}
+	docB, err := base64.StdEncoding.DecodeString(ok.Payload.Vulnerabilities)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to decode vulnerabilities: %v", err)
+	}
+	err = json.Unmarshal(docB, &doc)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshall vulnerabilities document: %v", err)
 	}
```
Comment (FrimIdan) [inline run/run.go]: need to move into a function that will return `doc` and `error`

File: run/run.go [hunk 12]
```diff
@@ -1,47 +1,154 @@
...
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get image commands : %v", err)
+	}
+
+	return &doc, commands, nil
+}
+
+func ExecuteLocalGrypeScan(imageName string, conf *config.Config) (*grype_models.Document, []*docker.FsLayerCommand, error) {
+	dbConfig := grype_db.Config{
+		DBRootDir:           "/tmp/",
```
Comment (FrimIdan) [inline run/run.go]: probably need to be configureable to support plugins that can't write to `/tmp/ - WDYT?

File: run/run.go [hunk 13]
```diff
@@ -1,47 +1,154 @@
...
+		InsecureUseHTTP:       conf.DockerConfig.InsecureRegistry,
+		Credentials: []anchore_image.RegistryCredentials{
+			{
+				Authority: "", // What is this?
+				Username:  conf.DockerConfig.User,
+				Password:  conf.DockerConfig.Password,
+				Token:     conf.DockerConfig.Token,
+			},
+		},
+	}
```
Comment (FrimIdan) [inline run/run.go]: duplicate - put into a function

File: run/run.go [hunk 14]
```diff
@@ -1,47 +1,154 @@
...
 	if err != nil {
-		log.Errorf("Failed to analyze using API: %s", err)
-	} else {
-		if !conf.JSONOutput {
-			log.Infof("Got results from Clair API")
-		}
+		return nil, fmt.Errorf("failed to parse name: %v", err)
+	}
+
+	err = image.Pull()
```
Comment (FrimIdan) [inline run/run.go]: so we basically "pull" the image twice, one in the scanning and one for the image commands?

**Review by FrimIdan — CHANGES_REQUESTED** [review]

File: run/run.go [hunk 15]
```diff
@@ -1,47 +1,154 @@
...
+	return &doc, commands, nil
+}
+
+func ExecuteLocalGrypeScan(imageName string, conf *config.Config) (*grype_models.Document, []*docker.FsLayerCommand, error) {
+	dbConfig := grype_db.Config{
+		DBRootDir:           "/tmp/",
+		ListingURL:          "https://toolbox-data.anchore.io/grype/databases/listing.json",
+		ValidateByHashOnGet: false,
+	}
+	provider, metadataProvider, dbStatus, err := grype.LoadVulnerabilityDB(dbConfig, true)
```
Comment (FrimIdan) [inline run/run.go]: should validate DB load with `func validateDBLoad(loadErr error, status *db.Status) error`
https://github.com/anchore/grype/blob/7e8ee40996ba3a4defb5e887ab0177d99cd0e663/cmd/root.go#L293

File: config/config.go (hunk 1 above)
Comment (akpsgit) [reply config/config.go]: replaced it with the existing severity definitions

File: run/run.go (hunk 10 above)
Comment (akpsgit) [reply run/run.go]: I removed the scan timeout. This should be very quick and the HTTP default timeout should be enough. WDYT?

Comment (FrimIdan) [reply run/run.go]: what is the `HTTP default timeout`?

File: forwarding/forwarding.go [hunk 16]
```diff
@@ -4,16 +4,16 @@ import (
...
 	"github.com/Portshift/klar/types"
+	grype_models "github.com/anchore/grype/grype/presenter/models"
 	log "github.com/sirupsen/logrus"
 	"io/ioutil"
 	"net/http"
 )
 
 type ImageVulnerabilities struct {
-	Vulnerabilities []*clair.Vulnerability   `json:"vulnerability"`
+	Vulnerabilities *grype_models.Document   `json:"vulnerability"`
```
Comment (fishkerez) [inline forwarding/forwarding.go:18]: why is the json annotaion is vulnerability and not vulnerabilities ?

File: main.go [hunk 17]
```diff
@@ -51,24 +50,20 @@ func main() {
...
 	if err != nil {
 		errMsg := fmt.Errorf("failed to execute scan: %w", err)
 		log.Error(errMsg)
 		result.ScanErr = types.ConvertError(errMsg)
 		exit(2, conf, result)
 	}
 
-	result.Vulnerabilities = filterVulnerabilities(conf.ClairOutput, vulnerabilities)
+	result.Vulnerabilities = filterVulnerabilities(conf.SeverityThreshold, vulnerabilities)
+	result.Vulnerabilities = vulnerabilities
```
Comment (fishkerez) [inline main.go]: why do you need this assignment?

File: run/run.go [hunk 18]
```diff
@@ -1,47 +1,154 @@
...
-		return nil, nil, fmt.Errorf("failed to pull pull fsLayers")
+	commands, err := GetImageCommands(conf)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get image commands : %v", err)
+	}
+
+	return &doc, commands, nil
+}
+
+func ExecuteLocalGrypeScan(imageName string, conf *config.Config) (*grype_models.Document, []*docker.FsLayerCommand, error) {
```
Comment (fishkerez) [inline run/run.go]: can you please add a comment above ExecuteLocalGrypeScan and ExecuteRemoteGrypeScan to describe shortly what each function does?

File: run/run.go (hunk 14 above)
Comment (akpsgit) [reply run/run.go]: It was always like this. it's not really pull of the tar, it's just the manifest. Didn't want to touch the commands logic part of this task.

File: run/run.go (hunk 10 above)
Comment (akpsgit) [reply run/run.go]: image

File: forwarding/forwarding.go (hunk 16 above)
Comment (akpsgit) [reply forwarding/forwarding.go:18]: didn't touch it, don't want to break stuff that might depend on it across the flow

File: config/config.go (hunk 2 above)
Comment (akpsgit) [reply config/config.go:62]: we can do it in a separate task, also move to urfav cli

File: run/run.go (hunk 10 above)
Comment (FrimIdan) [reply run/run.go]: please add scan timeout and do it configurable

**Review by FrimIdan — APPROVED** [review]

**Review by pbalogh-sa — APPROVED** [review]

**Pull request merged**